### PR TITLE
Add factory functions for FERemoteEvaluationCommunicator

### PR DESCRIPTION
**Pull request by jh66637**
@peterrum @fdrmrc @blaisb 

As requested in https://github.com/dealii/dealii/pull/16299#discussion_r1410262278, this PR introduces utility functions for the standard setup for point-to-point interpolation and Nitsche-type mortaring. In case of Nitsche-type mortaring, it is possible to hand over a NonMatching::MappingInfo object. In this case, the object is set up correctly in the function.

Most of the code was part of https://github.com/dealii/dealii/pull/16299 before switching to the use of the factory functions from this PR.

### CODE REVIEW

**Review by bangerth — COMMENTED** [review]
A small question.

Separately, would you mind (before merging) squashing your commits into one commit, or one commit for code/tests/changelog?

File: include/deal.II/matrix_free/fe_remote_evaluation.h [hunk 1]
```diff
@@ -434,6 +436,65 @@ class FERemoteEvaluationCommunicator : public Subscriptor
...
+
+  /**
+   * A factory function for the FERemoteEvaluationCommunicator in the case of
+   * Nitsche-type mortaring.
+   *
+   * @param[in] dof_handler Used DoFHandler object.
+   * @param[in] mapping Used Mapping object.
+   * @param[in] non_matching_faces_marked_vertices A vector of boundary face IDs
+   * that relate to non-matching interfaces. Each boundary face ID is
+   * accompanied by a lambda that specifies, where remote points can be found.
```
Comment (peterrum) [inline include/deal.II/matrix_free/fe_remote_evaluation.h:475]: "lambda that specifies, where remote points can be found" -> what does this mean?

File: include/deal.II/matrix_free/fe_remote_evaluation.h [hunk 2]
```diff
@@ -434,6 +436,65 @@ class FERemoteEvaluationCommunicator : public Subscriptor
 
 
 
+namespace Utilities
```
Comment (peterrum) [inline include/deal.II/matrix_free/fe_remote_evaluation.h:439]: Could you document the purpose of this namespace?

File: include/deal.II/matrix_free/fe_remote_evaluation.h [hunk 3]
```diff
@@ -963,6 +1024,316 @@ FERemoteEvaluationCommunicator<dim>::CopyInstructions::copy_data_entries(
...
+    auto remote_communicator =
+      std::make_unique<FERemoteEvaluationCommunicator<dim>>();
+
+    remote_communicator->reinit_faces(comm_objects,
+                                      face_batch_range,
+                                      global_quadrature_vector);
+
+    return remote_communicator;
+  }
+
```
Comment (peterrum) [inline include/deal.II/matrix_free/fe_remote_evaluation.h:1169]:
```suggestion



```

**Review by peterrum — COMMENTED** [review]
Few comments. I'll continue later.

**Comment by jh66637** [comment]
@peterrum The test needs https://github.com/dealii/dealii/pull/16359

**Comment by jh66637** [comment]
@peterrum @blaisb 
Since https://github.com/dealii/dealii/pull/16359 is merged we can proceed here. 
This is the last PR needed to finalize the tutorial on Nitsche-type mortaring :)

File: include/deal.II/matrix_free/fe_remote_evaluation.h [hunk 4]
```diff
@@ -999,6 +1073,373 @@ FERemoteEvaluationCommunicator<dim>::CopyInstructions::copy_data_entries(
...
+          tolerance, false, 0, marked_vertices);
+
+        // 2) Face batch IDs and number of faces in batch.
+        std::vector<std::pair<unsigned int, unsigned int>>
+          face_batch_id_n_faces;
+
+        // Points that are searched by rpe.
+        std::vector<Point<dim>> points;
+
+        // Temporarily setup FEFaceEvaluation to access the quadrature points at
```
Comment (kronbichler) [inline include/deal.II/matrix_free/fe_remote_evaluation.h:1140]:
```suggestion
        // Temporarily set up FEFaceEvaluation to access the quadrature points at
```
(you may want to look at all appearances, I'm going to stop highlighting them now.)

**Review by kronbichler — APPROVED** [review]

**Comment by jh66637** [comment]
@kronbichler Thank you for the review. I addressed all your comments and I think the documentation is more clear now :)

**Review by kronbichler — APPROVED** [review]

**Pull request merged**